### PR TITLE
Promote ArrayNode from a concept to a full subclass

### PR DESCRIPTION
**Pull request by arcondello**
Closes https://github.com/dwavesystems/dwave-optimization/issues/75 by implementing the virtual inheritance pattern.

### CODE REVIEW

**Comment by arcondello** [comment]
The failing tests highlight some of the downsides of virtual inheritance. We'd have to remove a lot of `constexpr` markers and similar.

File: dwave/optimization/include/dwave-optimization/graph.hpp [hunk 1]
```diff
@@ -293,21 +295,23 @@ NodeType* Graph::emplace_node(Args&&... args) {
...
+    auto uptr = std::make_unique<NodeType>(std::forward<Args&&>(args)...);
+    NodeType* ptr = uptr.get();
+
+    // Pass ownership of the lifespan to nodes_
+    nodes_.emplace_back(std::move(uptr));
 
     // Decisions get their topological index assigned immediately, in the order of insertion
     if constexpr (std::is_base_of_v<Decision, NodeType>) {
-        ptr->topological_index_ = num_decisions_++;
+        static_assert(std::is_base_of_v<DecisionNode, NodeType>);
```
Comment (wbernoudy) [inline dwave/optimization/include/dwave-optimization/graph.hpp:306]: This seems kind of redundant, no?

File: dwave/optimization/include/dwave-optimization/nodes/indexing.hpp [hunk 2]
```diff
@@ -141,20 +141,20 @@ class AdvancedIndexingNode : public Node, public Array {
...
-    BasicIndexingNode(Node* array_ptr, std::vector<slice_or_int> indices);
+    BasicIndexingNode(ArrayNode* array_ptr, std::vector<slice_or_int> indices);
 
     // Templated constructor to be used from C++
     // Just calls the runtime constructor for simplicity, but we could implement
     // a faster version by using more compile-time information
-    template <ArrayNode T, class... Indices>
-    explicit BasicIndexingNode(T* array_ptr, Indices... indices)
-            : BasicIndexingNode(static_cast<Node*>(array_ptr), make_indices(indices...)) {}
+    template <std::convertible_to<slice_or_int>... Indices>
```
Comment (wbernoudy) [inline dwave/optimization/include/dwave-optimization/nodes/indexing.hpp:155]: Cool

File: dwave/optimization/model.pxd [hunk 3]
```diff
@@ -141,13 +142,13 @@ cdef class Symbol:
 # extension base class, so to support that we assume all ArraySymbols are
 # also Symbols (probably a fair assumption)
 cdef class ArraySymbol(Symbol):
-    # Inheriting nodes must call this method from their __init__()
-    cdef void initialize_array(self, cppArray* array_ptr) noexcept
+    # Inheriting symbols must call this method from their __init__()
+    cdef void initialize_arraynode(self, Model model, cppArrayNode* array_ptr) noexcept
 
     # Hold Array* pointer. Again this is redundant, because we're also holding
```
Comment (wbernoudy) [inline dwave/optimization/model.pxd:148]: Should update this comment

File: dwave/optimization/src/nodes/mathematical.cpp [hunk 4]
```diff
@@ -40,12 +40,10 @@ struct BinaryOpNodeData : public NodeStateData {
 };
 
 template <class BinaryOp>
-BinaryOpNode<BinaryOp>::BinaryOpNode(Node* a_ptr, Node* b_ptr)
-        : Node(),
-          ArrayOutputMixin(broadcast_shape(dynamic_cast<const Array*>(a_ptr)->shape(),
-                                           dynamic_cast<const Array*>(b_ptr)->shape())),
```
Comment (wbernoudy) [inline dwave/optimization/src/nodes/mathematical.cpp:46]: Probably eliminating a bug here!

File: dwave/optimization/model.pxd [hunk 5]
```diff
@@ -19,7 +19,8 @@ from libcpp.memory cimport shared_ptr
 from libcpp.vector cimport vector
 
 from dwave.optimization.libcpp.array cimport Array as cppArray
```
Comment (wbernoudy) [inline dwave/optimization/model.pxd:21]: `cppArray` unused?

File: dwave/optimization/model.pxd [hunk 6]
```diff
@@ -141,13 +142,13 @@ cdef class Symbol:
...
-    # Inheriting nodes must call this method from their __init__()
-    cdef void initialize_array(self, cppArray* array_ptr) noexcept
+    # Inheriting symbols must call this method from their __init__()
+    cdef void initialize_arraynode(self, Model model, cppArrayNode* array_ptr) noexcept
 
     # Hold Array* pointer. Again this is redundant, because we're also holding
     # a pointer to Node* and we can theoretically dynamic cast each time.
     # But again, it's cheap and it simplifies things.
-    cdef cppArray* array_ptr
+    cdef cppArrayNode* array_ptr
```
Comment (wbernoudy) [inline dwave/optimization/model.pxd:151]: Should this have a different name to distinguish it from `cppArray*`?

File: tests/cpp/tests/test_nodes_mathematical.cpp [hunk 7]
```diff
@@ -404,32 +404,7 @@ TEMPLATE_TEST_CASE("NaryOpNode", "", functional::max<double>, functional::min<do
     }
 
     GIVEN("A vector<Node*> of constants") {
-        auto nodes = std::vector<Node*>{
```
Comment (wbernoudy) [inline tests/cpp/tests/test_nodes_mathematical.cpp:407]: Why is this removed?

**Review by wbernoudy — APPROVED** [review]
LGTM, comments are minor

File: dwave/optimization/include/dwave-optimization/graph.hpp (hunk 1 above)
Comment (arcondello) [reply dwave/optimization/include/dwave-optimization/graph.hpp:306]: It heads off someone accidentally inheriting from `Decision` but not `DecisionNode`. Like I did :laughing:

File: tests/cpp/tests/test_nodes_mathematical.cpp (hunk 7 above)
Comment (wbernoudy) [reply tests/cpp/tests/test_nodes_mathematical.cpp:407]: Oh I think I see, this constructor for `NaryOpNode` was removed, and the below tests covers the behavior.

File: dwave/optimization/include/dwave-optimization/graph.hpp (hunk 1 above)
Comment (wbernoudy) [reply dwave/optimization/include/dwave-optimization/graph.hpp:306]: Oops, missed the `Decision` vs `DecisionNode`! :+1:

File: dwave/optimization/model.pxd (hunk 6 above)
Comment (arcondello) [reply dwave/optimization/model.pxd:151]: Yeah, I thought about it, and erred on the side of brevity/readability. Also a lot fewer lines to change. At the Cython level I do think `ArrayNode` is more-or-less synonymous with `Array`.

File: dwave/optimization/model.pxd (hunk 5 above)
Comment (arcondello) [reply dwave/optimization/model.pxd:21]: Moved to the `.pyx`. Annoyingly I cannot just use `cppArrayNode` everywhere because Cython gets confused about inheritance and overloads.
